### DISCUSSION TITLE
Is there a way to use Rune with Send futures?

### DISCUSSION THREAD

**Discussion by elipp** [rune-rs/rune#473] [General]
Hi,

my main entry point has this weird requirement where I need:
- async recursion
- main entry point calls into a [insert scripting language] script, that should be able to potentially call that same entry point again.

I think the crux of the issue is here:

```
= help: within `Value`, the trait `Send` is not implemented for `NonNull<rune::runtime::shared::SharedBox<rune::runtime::Iterator>>`
note: future is not `Send` as this value is used across an await
  --> src/main.rs:53:48
   |
51 |     let mut vm = Vm::new(runtime, Arc::new(unit));
   |         ------ has type `Vm` which is not `Send`
52 |
53 |     let m = vm.async_call(["main"], (recurse,)).await.unwrap();
   |                                                ^^^^^^ await occurs here, with `mut vm` maybe used later`
```

I can fix it with using `tokio::LocalSet` and `#[async_recursion(?Send)]`, but just out of curiosity, I wanted to ask if there's a way to make this work :D I mean, seeing those `Arc`s all over the place kinda leads me to think it would be.

**Comment by udoprog** [comment]
What we support is `send_execute`: <https://docs.rs/rune/latest/rune/runtime/struct.Vm.html#method.send_execute>.

Example here: <https://github.com/rune-rs/rune/blob/main/examples/examples/tokio_spawn.rs>.

This also imposes limits on what can go into a future, and you have to clone the `Vm` for each call.

**Reply by elipp** [reply]
I see. Thank you for the answer!
